### PR TITLE
Improved pair plots for scenario discovery

### LINKED ISSUE #98: KDE/Contour plot option in PRIM show_pairs
Seaborn 0.11.0 (September 2020) added functionality to the `sns.pairplot()` function for kde (contours) and histograms (discretized heatmap). It would be great for EMA to add this functionality, especially since the contour plot would better show the distributions within PRIM results compared to the current scatter plot approach.

e.g. from https://seaborn.pydata.org/generated/seaborn.pairplot.html

### PR DESCRIPTION
**Pull request by steipatr**
These changes add some interesting plotting options to the pair plots for scenario discovery. New options include contour plots and bivariate histograms, as well as the option to leave the upper triangle empty to reduce visual clutter. There are no new required arguments, so existing code should still run, but will produce different figures. These changes are possible by moving the pair plot code from Seaborn's PairPlot to the more flexible PairGrid.

If accepted, this closes #98 .

Current EMA code.
download

Proposed new default behavior, with kernel density estimates on diagonal, contour plot and scatter plot. Less whitespace between plots and axes.
download

Some alternative options: cumulative density functions on diagonal, bivariate histogram in lower triangle, and empty upper triangle. Same whitespace between plots and axes as current implementation.
download

### CODE REVIEW

**Comment by EwoutH** [comment]
Thanks for this effort! I really dig the bivariate histograms. The contour plots look great for float and int inputs, but for categorical and bool inputs I find them confusing.

My suggestion for the default would be scatter plot on the upper triangle (like now), bivariate histograms on the lower one (new), and kernel density estimates on the diagonal (like now).

But maybe we can just add three arguments to the function, like

```Python
def show_pairs_scatter(upper=“scatter”, lower=“bivariate”, diag=“kde”):
```
And then we also allow `“contour”` for `upper` and `lower` and `“cumulative”` for `diag`.

Also really curious what @quaquel thinks!

**Comment by steipatr** [comment]
I agree that bivariate histograms are clearer across different variable types.

**Comment by quaquel** [comment]
I agree with the defaults suggested by @EwoutH and the offset suggested by @steipatr. Once those things are added, this is ready to be merged. Very good stuff!

**Comment by steipatr** [comment]
I have updated the default behavior as suggested by @EwoutH, and fiddled a bit with padding to make the boxes look better on top of categorical variables. Please note that bivariate histograms are considered experimental according to Seaborn docs. Current look: 
image

**Comment by quaquel** [comment]
Looks very good. I'll check the code asap and merge if everything is fine now.

File: ema_workbench/analysis/scenario_discovery_util.py [hunk 1]
```diff
@@ -357,59 +371,116 @@ def plot_pair_wise_scatter(x, y, boxlim, box_init, restricted_dims, cdf=False):
...
+            sns.kdeplot, fill=False, alpha=1, bw_adjust=1.2, levels=5, common_norm=False, cut=0
+        )
+    elif lower == "hist":
+        grid.map_lower(sns.histplot)
+    elif lower == "scatter":
+        grid.map_lower(sns.scatterplot)
+    elif lower == "none":
+        raise ValueError(f"Lower triangle cannot be none.")
+    else:
+        raise NotImplementedError(f"lower = {lower} not implemented.")
```
Comment (EwoutH) [inline ema_workbench/analysis/scenario_discovery_util.py:420]:
```suggestion
        raise NotImplementedError(f"lower = {lower} not implemented. Use either 'scatter', 'contour' or 'hist' (bivariate histogram) plots for lower triangle.")
```

**Review by EwoutH — COMMENTED** [review]
Thanks a lot for this effort! I will let the formal review to @quaquel, but I have a few minor suggestions.

File: ema_workbench/analysis/prim.py [hunk 2]
```diff
@@ -837,8 +839,18 @@ def show_pairs_scatter(self, i=None, dims=None, cdf=False):
         i : int, optional
         dims : list of str, optional
                dimensions to show, defaults to all restricted dimensions
-        cdf : bool, optional
-              plot diag as cdf or pdf
+        cdf : Boolean, optional
```
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/analysis/prim.py:842]: I agree, make this a string or enum (see https://github.com/quaquel/EMAworkbench/blob/90908518a77c3c03d86e1d78c150f2e6b39eb427/ema_workbench/analysis/plotting_util.py#L39)

**Comment by quaquel** [comment]
Thanks, @EwoutH for these suggestions. I agree with all of them and have added the easy ones already as commits. Only the enum/string thing still needs to be done.

File: ema_workbench/analysis/prim.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (steipatr) [reply ema_workbench/analysis/prim.py:842]: > Can we make this `diagonal` with possible values of `'kde'` and `'cdf'`? I think that would make it more consistent with how upper and lower is passed, and allows for adding a third option if we ever want to. And it's more explicit.

Yeah this makes total sense, I didn't want to deviate from

[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/analysis/prim.py:842]: Please make the modification (just strings is fine, but the enum would be more coherent with other parts of the workbench).

**Comment by quaquel** [comment]
I added the enum to the code. I guess this is ready to be merged, but @EwoutH if you have a chance to give it a last quick check, that would be great.

**Comment by EwoutH** [comment]
Personally I find ENUM a bit overkill for such a construct, but it can't hurt either.

Did some manual testing happen?

If nothing major has changed since my last review, it's good to go.

**Comment by quaquel** [comment]
I agree on the enum comment. I decided to go down the enum route for parallelism with density plots. I would even argue that the density plots might also be overkill (4 options only). I have tested the code with the prim example. No further changes took place.

**Comment by quaquel** [comment]
Thanks @steipatr

**Comment by EwoutH** [comment]
Thanks a lot Patrick!

**Comment by steipatr** [comment]
Thank you both for your efforts! I think it's a great enhancement. Now to generalize it to multiple subspaces...

**Pull request merged**